### DISCUSSION TITLE
Generating hashes of ReadOnlySpan<char> that are equivalent to string.GetHashCode()

### DISCUSSION THREAD

**Discussion by PatVax** [dotnet/runtime#102721] [Q&A]
I am writing a Specialized Dictionary class for string values where the lookup is done by a char sequence. This could be any string, `ReadOnlySpan<char>`, `ReadOnlyMemory<char>`, `ReadOnlySequence<char>`. The goal is to retrieve a string that is equivalent to the input without allocations or create it when it doesn't exist. I know that we have a `ReadOnlySpan<char>` overload of `string.GetHashCode` but none for `ReadOnlySequence<char>`. 

To make it work with `ReadOnlySequence<char>` there are following possibilities I could think of so far:

- Check if the sequence is made of a single segment and if so use the `FirstSpan` property with the `GetHashCode` overload or if it is made of multiple segments copy the data into a stack allocated buffer and pass this span into `GetHashCode`. This is almost optimal. Only in the case of multiple segments we need a seemingly unnecessary data copy.
- Use the `HashCode` struct to produce hashes in every case. In this case I am not sure if the resulting hashes would be as good as the ones produced in `string.GetHashCode` with the Marvin algorithm. It would look something like this:
```csharp
int Hash(ReadOnlySequence<char> seq)
{
    HashCode hash = new();
    foreach (var memory in seq)
    foreach (var c in memory)
        hash.Add(c);
    return hash.ToHashCode();
}

int Hash(ReadOnlySpan<char> span)
{
    HashCode hash = new();
    foreach (var c in span)
        hash.Add(c);
    return hash.ToHashCode();
}

int Hash(string s) => Hash(s.AsSpan());
int Hash(ReadOnlyMemory<char> memory) => Hash(memory.Span);
```
- Roll out my own hashing algorithm like Murmur2 as compared in [this stackoverflow answer](https://softwareengineering.stackexchange.com/a/145633/373531) and use it for all cases.

So what I am looking for is a .NET built in way to produce `string.GetHashCode` equivalent hash from `ReadOnlySequence<char>` without making a copy of the sequence when it is made of multiple segments.

**Comment by EgorBo** [comment]
Consider using XxHash3, we might even use it as a default String.GetHashCode algorithm in future (https://github.com/dotnet/runtime/issues/85206)

**Comment by PatVax** [comment]
I just found XxHash in System.IO.Hashing namespace. Is this the right thing to use?

**Comment by bartonjs** [comment]
If you want the same hashes as are produced by System.String; I believe they get exposed via [`CompareInfo.GetHashCode(ReadOnlySpan<char>, CompareOptions)`](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.gethashcode?view=net-8.0#system-globalization-compareinfo-gethashcode(system-readonlyspan((system-char))-system-globalization-compareoptions))

**Reply by PatVax** [reply]
> If you want the same hashes as are produced by System.String; I believe they get exposed via [`CompareInfo.GetHashCode(ReadOnlySpan<char>, CompareOptions)`](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.gethashcode?view=net-8.0#system-globalization-compareinfo-gethashcode(system-readonlyspan((system-char))-system-globalization-compareoptions))

Yes I got that far. I think the static version of `string.GetHashCode` is essentially the same as `CompareInfo.GetHashCode(ReadOnlySpan<char>, CompareOptions)`. This is fine. I was looking for an overload with `CompareInfo.GetHashCode(ReadOnlySequence<char>, CompareOptions)` my guess is that it simply doesn't exist. For know I am going to use XxHash3 from System.IO.Hashing for anything that could be considered a string. If sometime in the future `ReadOnlySequence<char>` overloads are built in I will gladly switch back.

**Reply by TahirAhmadov** [reply]
This has been helpful to me. I just put the 1033 `CompareInfo` into a static readonly field, and use it for both the strings and the spans, and it works great. Thanks.
